### DISCUSSION TITLE
Emote Tab Completion Strategy

### DISCUSSION THREAD

**Discussion by leon-richardt** [Chatterino/chatterino2#4414] [Ideas]
# Description

Originally [suggested](https://discord.com/channels/253897423439659008/459650160658481152/1079155422684008458) by `neuro#3365` on the Chatterino Discord:
> Random user feedback: A channel has both "Looking" and "BorpaLookingAtYou". I'm guessing tab complete goes alphabetically, because no matter what you type (even wholly typing 'Looking'), it always goes to BorpaLookingAtYou first. Seems like it should go for the more exact matches first, rather than just alphabetically

> Thinking about it, instead of alphabetical just do length, shortest to longest. The shortest match is going to be the closest match 100% of the time anyways. Length ties are gonna be so rare there's no point even thinking about them

---

I assume by "shortest match" the user means the specific emote whose name is covered most by the pattern.

In other words, let $P$ be the pattern that is being tab-completed and let $T_1, \ldots, T_n$ be the emote candidates, i.e. $P$ is a case-insensitive substring of $T_i$ for all $i = 1, \ldots, n$. Order the candidates (in descending order) by the values $\frac{|P|}{|T_i|}$ (where $|S|$ denotes the length of string $S$). This is the order tab completion will cycle through.

Intuitively, this makes sure that emotes that are "covered more" by the search pattern appear before emotes that are "covered less". I agree with the user that this is likely to be a better approach when compared to strict alphabetical order.

**Note:** This also implies that exact matches (disregarding case) will always come first in completion order. (A behavior I have wanted for a long time!) 

# Examples

## Example 1
Consider $P$ = `dank` and $T_1$ = `DankG`, $T_2$ = `EVENDANKERTHANWAYTOODANK`, $T_3$ = `FeelsDankMan`, $T_4$ = `pajaDank` (given in the current alphabetical ordering). Calculating the "cover ratios" yields:

- $\frac{|P|}{|T_1|} = \frac{4}{5} = 0.8$
- $\frac{|P|}{|T_2|} = \frac{4}{24} \approx 0.167$
- $\frac{|P|}{|T_3|} = \frac{4}{12} \approx 0.333$
- $\frac{|P|}{|T_4|} = \frac{4}{8} = 0.5$

Thus, the tab completion order would be `DankG` → `pajaDank` → `FeelsDankMan` → `EVENDANKERTHANWAYTOODANK`.

## Example 2
`neuro#3665` suggests that "length ties are gonna be [...] rare" but I'm not sure I agree. One can imagine scenarios where $P$ is chosen in a manner that allows for more candidates with the same length.

For example, consider $P$ = `forsen` and $T_1$ = `forsen1`, $T_2$ = `forsen2`, $T_3$ = `forsen3`, $T_4$ = `forsen4`, $T_5$ = `forsenC`, $T_6$ = `forsenD`, $T_7$ = `forsenE`. This pattern is not unusual among Twitch emotes. In this case, a tie break is needed (although I'm not sure it's possible to be much cleverer than alphabetical order).

However, this extreme case only occurs if $P$ is chosen in a way that admits many candidates. Choosing $P$ in a "more specific" manner would have prevented a long candidate list.

# Discussion
What does everyone think about implementing this idea for emote completion order as opposed to the current alphabetical order? Personally, I think it is a more ergonomic approach to tab completion and would be in favor of implementing it.

Can you think of any edge cases where the described change would lead to a worse UX than the current alphabetical order?

**Comment by TranRed** [comment]
Sounds like a fun idea to explore.

My first idea for a tiebreaker mechanic Example 2 would be a two step approach:
1. position of the match (ascending)
2. alphabetical order (ascending)

**Reply by leon-richardt** [reply]
> position of the match (ascending)

Very interesting! This might even allow us to ditch the option `Only search for emote autocompletion at the start of emote names` since those would always be at the start of the completion order anyways, correct?

**Reply by TranRed** [reply]
It's not 100% the same, right? Unless that priority is higher than cover ratio.

with the tie-breaker it would just be a thing in `forsen1` vs. `1forsen`

but `1forsen` would still prioritize over `ForsenLookingAtYou`, because it still has a higher cover ratio

**Reply by leon-richardt** [reply]
Another factor to consider for tie breaking is case-sensitivity.

**Comment by TranRed** [comment]
Another thing I just observed: (I hope that's not a problem of me being on a slighty older version right now)

should the :-menu obey the same order as tabbing? 
Currently it does not for me. 
I would personally prefer consistency between those features. 

Order in :-menu 
![grafik](https://user-images.githubusercontent.com/16636423/221412714-0564ed15-8e5b-4808-b487-bd5cf27d04dc.png)

where as tab order for me currently is alphabetically: forsen, forsen1, forsen2, forsen3, forsen4, forsenA, forsenAngry, forsenBased (...)

**Reply by Mm2PL** [reply]
I'm on "Chatterino Nightly + Dankerino patches 2.4.1 (commit [3d2b32c7](https://github.com/Mm2PL/dankerino/commit/3d2b32c7); based on [4923549f](https://github.com/chatterino/chatterino2/commit/4923549f) modified) built on 2023-02-24 with Qt 5.15.8
Running on Arch Linux, kernel: 6.1.10-arch1-1" and it doesn't have the same order

Bit off topic but i think we should create a unified completion system that takes both :-search and tab completion into one thing under the hood, it should probably also obey the new completion rules

**Reply by TranRed** [reply]
> Bit off topic but i think we should create a unified completion system that takes both :-search and tab completion into one thing under the hood

Yeah that's the chance I am seeing here when the topic of tab order gets adressed/changed anyways with this.
Two birds with one stone

**Comment by dnsge** [comment]
I was just thinking about this the other day and could have sworn that I had already had a PR merged that fixed this. Just went and checked, PR #1895 implemented putting exact matches at the top of the `InputCompletionPopup` but didn't make any changes to the tab autocompletion order. 

It would probably be best to unify the tab autocompletion and `InputCompletionPopup` suggestion order logic (and maybe the same thing with autocompleting `@username` suggestions, as I recall trying to implement a feature that would have benefitted from that).

**Comment by astradamus** [comment]
neuro here. Damn! Cool to see this little offhand thought bloom into a full on idea!

> neuro#3665 suggests that "length ties are gonna be [...] rare" but I'm not sure I agree. One can imagine scenarios where is chosen in a manner that allows for more candidates with the same length.

Yes, this is true, I should instead have said "are going to be rare for a savvy user". Like if I want to get forsenE, I'm not gonna type forsen->tab, I'm gonna type sene->tab, or whatever other minimal unique segment occurs to me (ideally left-hand only).

So how that particular situation is resolved is kinda irrelevant because the user can still control whether or not they get into it. Whereas in the original example of Looking vs. BorpaLookingAtYou, there is literally no possible input you can do to exclude the longer version from the search.

**Reply by astradamus** [reply]
If you guys want to really get nuts, the full evolved form of this feature is to account for case in the target, the way Jetbrains IDEs do autocompletion.

_BLA->tab_ or _BLAY->tab_ or _blay->tab_ would find BorpaLookingAtYou

This might be overkill, but I definitely constantly wish it was there whenever I'm typing a PascalCase emote :P

**Comment by robmonte** [comment]
I definitely agree that it should find exact matches first. As a user I find it very jarring when I am looking for an emote that I can't quite remember the name of but know the first word of it, and upon pressing Tab it _completely_ changes my input. At the very least, having a toggle for this behavior would be extremely welcome.

Secondly, I think a visualization for Tab-completion would be really helpful. This could also be toggleable behavior, where you can enable Tab-completion to display the same pop-up that Colon-completion does. (Wasn't sure if I should make this a separate discussion post or add it here).

On that note of the pop-up visualization, why is it so tiny? On my 150% scaled 4K monitor, it shows 4 emote lines at a time, except I can barely see the emotes/read the text because it is so tiny. On my 100% scaled monitor, I can see the emote/read the text just fine, except the window is so short that it only shows 2 emote lines at a time. I'd argue even the 4 is too little and uncomfortable to use.

**Reply by Nerixyz** [reply]
> On that note of the pop-up visualization, why is it so tiny? On my 150% scaled 4K monitor, it shows 4 emote lines at a time, except I can barely see the emotes/read the text because it is so tiny. On my 100% scaled monitor, I can see the emote/read the text just fine, except the window is so short that it only shows 2 emote lines at a time.

Regarding the font-size: This sounds like something #3690 could/should address or has already addressed.

**Comment by LosFarmosCTL** [comment]
> and upon pressing Tab it completely changes my input

I think you might have disabled 'Only search for emote autocompletion at the start of emote names' in settings, it definitely should not change the entire input, tabbing by default only looks for emotes that start with whatever you typed.

**Reply by robmonte** [reply]
Well not quite, I am simply agreeing with the original suggestion from neuro. I don't want _only_ emotes that begin with my prefix to be suggested (which is what that setting does). I'm saying when I type in something and press Tab, it is jarring when I do expect the _first_ suggestions to be emotes that begins with exactly what I typed. But other emotes containing the substring I entered should still be in the list, because sometimes I mash Tab looking for a specific one that I only partially remember.
